src/api-service/__app__/onefuzzlib/azure/nsg.py

```diff
@@ -0,0 +1,480 @@
+    network_client = get_network_client()
+    try:
+        nsg = network_client.network_security_groups.get(resource_group, name)
+        return cast(NetworkSecurityGroup, nsg)
+    except (ResourceNotFoundError, CloudError) as err:
+        logging.debug("nsg %s does not exist: %s", name, err)
+        return None
+
+
+def create_nsg(name: str, location: Region) -> Union[None, Error]:
+    resource_group = get_base_resource_group()
+
+    logging.info("creating nsg %s:%s:%s", resource_group, location, name)
+    network_client = get_network_client()
+
+    params: Dict = {
+        "location": location,
+    }
+
+    if "ONEFUZZ_OWNER" in os.environ:
+        params["tags"] = {"OWNER": os.environ["ONEFUZZ_OWNER"]}
+
+    try:
+        network_client.network_security_groups.begin_create_or_update(
+            resource_group, name, params
+        )
+    except (ResourceNotFoundError, CloudError) as err:
+        if is_concurrent_request_error(str(err)):
+            logging.debug(
+                "create NSG had conflicts with concurrent request, ignoring %s", err
+            )
+            return None
+        return Error(
+            code=ErrorCode.UNABLE_TO_CREATE,
+            errors=["Unable to create nsg %s due to %s" % (name, err)],
+        )
+    return None
+
+
+def list_nsgs() -> List[NetworkSecurityGroup]:
+    resource_group = get_base_resource_group()
+    network_client = get_network_client()
+    return list(network_client.network_security_groups.list(resource_group))
+
+
+def update_nsg(nsg: NetworkSecurityGroup) -> Union[None, Error]:
+    resource_group = get_base_resource_group()
+
+    logging.info("updating nsg %s:%s:%s", resource_group, nsg.location, nsg.name)
+    network_client = get_network_client()
+
+    try:
+        network_client.network_security_groups.begin_create_or_update(
+            resource_group, nsg.name, nsg
+        )
+    except (ResourceNotFoundError, CloudError) as err:
+        if is_concurrent_request_error(str(err)):
+            logging.debug(
+                "create NSG had conflicts with concurrent request, ignoring %s", err
+            )
+            return None
+        return Error(
+            code=ErrorCode.UNABLE_TO_CREATE,
+            errors=["Unable to update nsg %s due to %s" % (nsg.name, err)],
+        )
+    return None
+
+
+def ok_to_delete(active_regions: Set[Region], nsg_region: str, nsg_name: str) -> bool:
+    return nsg_region not in active_regions and nsg_region == nsg_name
+
+
+def is_one_fuzz_nsg(nsg_region: str, nsg_name: str) -> bool:
+    return nsg_region == nsg_name
+
+
+def delete_nsg(name: str) -> bool:
+    # NSG can be only deleted if no other resource is associated with it
+    resource_group = get_base_resource_group()
+
+    logging.info("deleting nsg: %s %s", resource_group, name)
+    network_client = get_network_client()
+
+    try:
+        network_client.network_security_groups.begin_delete(resource_group, name)
+        return True
```

Comment:
it looks we need the result of the operation here to confirm the deletion 
```suggestion
        delete_nsg = network_client.network_security_groups.begin_delete(resource_group, name)
        return delete_nsg.done()
```

Reply:
When we call begin_delete we get Poller object, that if we call "done()" most likely will return false, since it is along running operation and deletion is in progress. [Poller object "done" documentation](https://docs.microsoft.com/en-us/python/api/azure-core/azure.core.polling.lropoller?view=azure-python#done--)

So we need to wait for it to be done, I don't think we want to block the execution of the azure function to wait for deletion to complete.

The behavior in the PR aligns with the current implementation across OneFuzz azure module. For example:
[vmss](https://github.com/microsoft/onefuzz/blob/6cdb5e17dbcd7b0fca2e06580842083c7a8fe419/src/api-service/__app__/onefuzzlib/azure/vmss.py#L69)

[vm](https://github.com/microsoft/onefuzz/blob/6cdb5e17dbcd7b0fca2e06580842083c7a8fe419/src/api-service/__app__/onefuzzlib/azure/vm.py#L160)

The  check for whether nsg deletion call should retried happens in parallel inside of `timer_proxy`. And this behavior is aligned with existing behavior from across OneFuzz azure module

Reply:
after discussion and code diving. I update my suggestion to change the name of the function to `begin_delete` to comunicate that a `true` return value does not necessarily mens that the resource was deleted